src/v/v8_engine/script_dispatcher.h

```diff
@@ -0,0 +1,151 @@
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "bytes/iobuf.h"
+#include "model/metadata.h"
+#include "outcome.h"
+#include "v8_engine/errc.h"
+#include "v8_engine/script.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/loop.hh>
+#include <seastar/core/shared_ptr.hh>
+
+#include <absl/container/node_hash_map.h>
+
+#include <exception>
+#include <iostream>
+#include <variant>
+
+namespace v8_engine {
+
+struct script_compile_params {
+    model::topic_namespace topic;
+    ss::sstring function_name;
+    iobuf code;
+    size_t max_heap_size_bytes{10000};
+    size_t run_timeout_ms{200};
+};
+
+struct script_remove_params {
+    model::topic_namespace topic;
+};
+
+// This class implement database for v8_runtimes. It can be updated by
+// data_policy_manager. Fetch request will only run get for getting v8_runtimes
+template<typename Executor>
+class script_dispatcher {
+    constexpr static size_t _mutex_count{50};
+
+public:
+    explicit script_dispatcher(Executor& executor)
+      : _executor(executor) {}
+
+    ss::future<> start() {
+        (void)ss::with_gate(_gate, [this] {
+            return ss::do_until(
+              [this] { return _as.abort_requested(); },
+              [this]() -> ss::future<> { co_await update_runtimes(); });
+        });
+
+        co_return;
+    }
+
+    ss::future<> stop() {
+        _as.request_abort();
+        co_await _gate.close();
+    }
+
+    void update(
+      model::topic_namespace topic, ss::sstring function_name, iobuf code) {
+        _event_queue.emplace_back(script_compile_params{
+          .topic = std::move(topic),
+          .function_name = std::move(function_name),
+          .code = std::move(code)});
+    }
+
+    void remove(model::topic_namespace topic) {
+        _event_queue.emplace_back(
+          script_remove_params{.topic = std::move(topic)});
+    }
+
+    /// This method return v8_runtimes for data-policy (aka topic on curent
+    /// moment). If hash_table contains compile params get needs to compile this
+    /// script, so on this moment we need to lock hash_table. Also if sbmd try
+    /// to get v8_runtimes it needs to wait compilation and do not compile twice
+    result<ss::lw_shared_ptr<script>, errc>
+    get(const model::topic_namespace& topic) {
+        auto it = _runtimes.find(topic);
+        if (it == _runtimes.end()) {
+            return errc::v8_runtimes_not_exist;
+        }
+
+        if (it->second.get() == nullptr) {
+            return errc::v8_runtimes_compiling;
+        }
+
+        return it->second;
+    }
+
+private:
+    // Internal function which read updates from queue and change hash_map with
+    // v8_runtimes. TODO: add logging
+    ss::future<> update_runtimes() {
```

Comment:
This approach does not allow lazy compilation. So each core will contain v8_runtime. I do not know, is it good idea?

Reply:
![image](https://media.giphy.com/media/YNxE1XvNgNCoVqEFa4/giphy.gif)

Reply:
You can have lazy compilation without each core containing a v8 runtime right? You can invoke the compilation command on the v8_runtimes core , and the future resolves when the command has finished. Multiple calls can also wait on this event to complete as well:
```
std::map<script_name, std::vector<ss::promise<>> _updates;
```
Resolve all of the promises when the compile is done.

Reply:
Why do you use multiply promise for each script_name.

My first idea: on each time we have one core which home for ntp, so clients will read from that core. We will update map for each core, this map contains compile_params or v8_runtimes. Also we will subscribe got changing home core, and delete runtimes from prev home core.

Maybe in my current code I can put updates to queue on home core

Reply:
Multiple promises because there is a promise its for every caller for a given script:
```
std::map<int, std::vector<ss::promise<>> _alerts;
ss::future<> alert_me(int key){
  auto [itr, success] = _alerts.emplace(key);
  if(!success){
    return ss::now();
  }
  itr->second.emplace_back();
  return itr->second.get_future();
}
```